rmrk_proxy/proxy.rs

```diff
@@ -0,0 +1,490 @@
+// Copyright (c) 2023 Astar Network
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the"Software"),
+// to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#[ink::contract]
+mod rmrk_proxy {
+    use crate::{
+        ensure,
+        ProxyError,
+        Result,
+    };
+    use ink::{
+        env::{
+            call::{
+                build_call,
+                ExecutionInput,
+                Selector,
+            },
+            hash,
+            DefaultEnvironment,
+        },
+        prelude::vec::Vec,
+    };
+    use openbrush::{
+        contracts::{
+            ownable::*,
+            psp34::Id,
+            reentrancy_guard::*,
+        },
+        modifiers,
+        traits::Storage,
+    };
+
+    // Proxy contract storage
+    #[ink(storage)]
+    #[derive(Default, Storage)]
+    pub struct RmrkProxy {
+        #[storage_field]
+        ownable: ownable::Data,
+        #[storage_field]
+        guard: reentrancy_guard::Data,
+        #[storage_field]
+        proxy: crate::types::Data,
+    }
+
+    impl RmrkProxy {
+        #[ink(constructor)]
+        pub fn new(
+            rmrk_contract: AccountId,
+            catalog_contract: AccountId,
+            mint_price: Balance,
+        ) -> Self {
+            let mut instance = Self::default();
+            instance.proxy.rmrk_contract = Option::Some(rmrk_contract);
+            instance.proxy.catalog_contract = Option::Some(catalog_contract);
```

Comment:
You never use None in the contract. Is Option needed?
You are doing a lot of unwrapping due to this

Reply:
Option was a workaround for the following error which I couldn't solve
<img width="976" alt="image" src="https://github.com/swanky-dapps/rmrk-lazy-mint/assets/8452361/d022d8e0-e79b-4a33-a2c4-e2cf3d3deb6a">